### PR TITLE
mitosheet: standardize column toggle all

### PR DESCRIPTION
**Pull request by naterush**
# Description

1. Closes https://github.com/mito-ds/monorepo/issues/598
2. Introduces a MultiToggleColumns component - which is _sooo_ much cleaner!

Also opened this issue: https://github.com/mito-ds/monorepo/issues/625 (which is unrelated to this PR)

# Testing

Test all the toggles that are changes. I did!

# Documentation

No.

### CODE REVIEW

File: mitosheet/src/components/elements/MultiToggleColumns.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                const newSelectedColumnIDs = newSelectedIndexes.map(index => {return columnIDs[index]});
+                props.onChange(newSelectedColumnIDs);
+            }}
+            height='medium'
+        >
+            {columnIDsAndDtype.map(([columnID, columnDtype], index) => {
+                const columnHeader = columnIDsMap[columnID];
+
+                // We turn off and disable the toggle in the case it is included in the id variables, 
+                // as pandas automatically filters the id variables out from the value variables
```
Comment (naterush) [inline mitosheet/src/components/elements/MultiToggleColumns.tsx:43]: Cleanup

File: mitosheet/src/components/elements/MultiToggleColumns.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                const toggle = props.selectedColumnIDs.includes(columnID);
+                const disabled = props.disabledColumnIDs !== undefined && props.disabledColumnIDs.includes(columnID);
```
Comment (naterush) [inline mitosheet/src/components/elements/MultiToggleColumns.tsx:45]: Add a function that you can optionally pass

File: mitosheet/src/components/taskpanes/FillNa/FillNaTaskpane.tsx [hunk 3]
```diff
@@ -212,32 +187,24 @@ const FillNaTaskpane = (props: FillNaTaskpaneProps): JSX.Element => {
...
-                >
-                    {Object.entries(columnDtypeMap).map(([columnID, columnDtype], index) => {
-                        const columnHeader = columnIDsMap[columnID];
-                        const toggle = params.column_ids.includes(columnID);
-                        const disabled = (params.fill_method.type === 'mean' || params.fill_method.type === 'median') && 
+                <MultiToggleColumns
+                    sheetData={sheetData}
+                    selectedColumnIDs={params.column_ids}
+                    disabledColumnIDs={[]} // TODO
+                    /**
```
Comment (naterush) [inline mitosheet/src/components/taskpanes/FillNa/FillNaTaskpane.tsx:194]: TODO

File: mitosheet/src/components/taskpanes/Merge/MergeTaskpane.tsx [hunk 4]
```diff
@@ -206,61 +202,21 @@ const MergeTaskpane = (props: MergeTaskpaneProps): JSX.Element => {
...
-                                } else {
-                                    columnIDs.forEach((columnID) => {
-                                        removeIfPresent(newSelectedColumnIDsOne, columnID);
-                                    })
-                                }
-
+                    <MultiToggleColumns
+                        sheetData={sheetDataOne}
+                        selectedColumnIDs={params.selected_column_ids_one.concat(params.merge_key_column_ids.map(([one, ]) => {return one}))}
+                        /**
```
Comment (naterush) [inline mitosheet/src/components/taskpanes/Merge/MergeTaskpane.tsx:208]: Cleanup

File: mitosheet/src/components/elements/MultiToggleColumns.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    selectedColumnIDs: ColumnID[],
+    onChange: (newSelectedColumnIDs: ColumnID[]) => void;
+    height?: Height;
+    disabledColumnIDs?: ColumnID[],
+    getIsDisabledColumnID?: (columnID: ColumnID, columnHeader: ColumnHeader, columnDtype: string) => boolean;
+    getDisplayColumnHeaderOverride?: (columnID: ColumnID, columnHeader: ColumnHeader) => string;
+}
+
+/**
+ * The MultiToggleDataframe component allows the user to toggle and select multiple
```
Comment (aarondr77) [inline mitosheet/src/components/elements/MultiToggleColumns.tsx:22]: The name is MultiToggleColumns

File: mitosheet/src/components/import/XLSXImportConfigScreen.tsx [hunk 6]
```diff
@@ -143,15 +143,10 @@ function XLSXImportConfigScreen(props: XLSXImportConfigScreenProps): JSX.Element
                             searchable
                             height='medium'
                             emptyMessage='There are no sheets to choose from. Either the workbook is empty or password protected.'
-                            toggleAllIndexes={(indexesToToggle) => {
+                            onToggleAll={(newSelectedIndexes) => {
                                 // TODO: this is broken. The toggle all is not working properly!
```
Comment (aarondr77) [inline mitosheet/src/components/import/XLSXImportConfigScreen.tsx:147]: Remove this TODO

File: mitosheet/src/components/elements/MultiToggleColumns.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                const newSelectedColumnIDs = newSelectedIndexes.map(index => {return columnIDs[index]});
+                props.onChange(newSelectedColumnIDs);
+            }}
+            height='medium'
+        >
+            {columnIDsAndDtype.map(([columnID, columnDtype], index) => {
+                const columnHeader = columnIDsMap[columnID];
+
+                // We turn off and disable the toggle in the case it is included in the id variables, 
+                // as pandas automatically filters the id variables out from the value variables
```
Comment (aarondr77) [inline mitosheet/src/components/elements/MultiToggleColumns.tsx:44]: Does this comment apply here? I think its specific to the melt taskpane

File: mitosheet/src/components/elements/MultiToggleBox.tsx [hunk 8]
```diff
@@ -260,7 +260,12 @@ const MultiToggleBox = (props: {
                             if (props.disabled) {
                                 return;
                             }
-                            toggleAllIndexes(nonDisabledDisplayedIndexes, !displayedNonDisabledAllToggled)
+
+                            if (!displayedNonDisabledAllToggled) {
```
Comment (aarondr77) [inline mitosheet/src/components/elements/MultiToggleBox.tsx:264]: Is there a better name for this `displayedNonDisabledAllToggled`? It's quite hard to understand what this is doing. 

Or could you at least add a comment explaining the different parts. 

Displayed = not removed from the search
Non Disabled = not disabled 
All Toggled = The state of the toggle all button?

**Comment by aarondr77** [comment]
Test is having the same errors as the others ones. Once its fixed up LGTM.

**Review by aarondr77 — COMMENTED** [review]
Test is having the same errors as the others, once their fixed up LGTM

File: mitosheet/src/components/elements/MultiToggleBox.tsx (hunk 8 above)
Comment (naterush) [reply mitosheet/src/components/elements/MultiToggleBox.tsx]: gave it a much better name, it was confusing

**Pull request merged**